airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/record.py

```diff
@@ -0,0 +1,118 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+
+from io import TextIOWrapper
+from json import loads
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Union
+
+from .query import ShopifyBulkQuery
+from .tools import END_OF_FILE, BulkTools
+
+
+class ShopifyBulkRecord:
+    def __init__(self, query: ShopifyBulkQuery) -> None:
+        self.composition = query.record_composition
+        self.record_process_components = query.record_process_components
+        self.components: List[str] = self.composition.get("record_components", []) if self.composition else []
+        self.buffer: List[MutableMapping[str, Any]] = []
+        self.tools: BulkTools = BulkTools()
+
+    @staticmethod
+    def check_type(record: Mapping[str, Any], types: Union[List[str], str]) -> bool:
+        record_type = record.get("__typename")
+        if isinstance(types, list):
+            return any(record_type == t for t in types)
+        else:
+            return record_type == types
+
+    def record_new(self, record: MutableMapping[str, Any]) -> None:
+        record = self.component_prepare(record)
+        record.pop("__typename")
+        self.buffer.append(record)
+
+    def record_new_component(self, record: MutableMapping[str, Any]) -> None:
+        component = record.get("__typename")
+        record.pop("__typename")
+        # add component to its placeholder in the components list
+        self.buffer[-1]["record_components"][component].append(record)
+
+    def component_prepare(self, record: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        if self.components:
+            record["record_components"] = {}
+            for component in self.components:
+                record["record_components"][component] = []
+        return record
+
+    def buffer_flush(self) -> Iterable[Mapping[str, Any]]:
+        if len(self.buffer) > 0:
+            for record in self.buffer:
+                # resolve id from `str` to `int`
+                record = self.record_resolve_id(record)
+                # process record components
+                yield from self.record_process_components(record)
+            # clean the buffer
+            self.buffer.clear()
+
+    def record_compose(self, record: Mapping[str, Any]) -> Optional[Iterable[MutableMapping[str, Any]]]:
+        """
+        Step 1: register the new record by it's `__typename`
+        Step 2: check for `components` by their `__typename` and add to the placeholder
+        Step 3: repeat until the `<END_OF_FILE>`.
+        """
+
+        if self.check_type(record, self.composition.get("new_record")):
+            # emit from previous iteration, if present
+            yield from self.buffer_flush()
+            # register the record
+            self.record_new(record)
+        # components check
+        elif self.check_type(record, self.components):
```

Comment:
No, since the record could be partially returned, there is no reason to raise an error for the record that is missing some `line_item` element, or other record component, as for me. WDYT?

Reply:
why would a record be partially returned?

Reply:
For instance:
- Fulfillment Orders stream record is:
for each `fulfullment_order` there are 2 record components:
- Line Item (List of dict)
- Merchant Request (List of dict)

Full record with components looks like:
```
{
  "id": 123, 
   ...., 
   ...., 
   "line_items": [
       {
          "id": 444,
           ....,
           <some_more_properties>,
           ....,
       },
       {....},
       {....}
    ], 
   "merchant_requests": [
       {
          "id": 555,
           ....,
           <some_more_properties>,
           ....,
       },
       {....},
       {....}
    ]
}
```

Partial record looks like:
```
{
  "id": 123, 
   ...., 
   ...., 
   "line_items": [],  
   "merchant_requests": [],
}
```

With the given information, why should we skip the `fullfilment_order` record, that doesn't have any of : [LineItem, MerchantRequest] ?

Do you recommend skipping the record based on that? 
This will cause `missing records`, if so.

Reply:
Additionaly, the basic example is here:  
https://shopify.dev/docs/api/usage/bulk-operations/queries#the-jsonl-data-format

Where: 
```
{"id":"gid://shopify/Product/1921569226808",  ...., "some_other_important_fields": ....,}
{"id":"gid://shopify/ProductVariant/19435458986123","title":"52","__parentId":"gid://shopify/Product/1921569226808"}
```
`Product` - the main record
`ProductVariant` - is the record component that  should be merged with the main one, if it's the case

Like:

```
{
"id":"gid://shopify/Product/1921569226808",  
...., 
"some_other_important_fields": ...., 
product_variants: [
   {
    "id":"gid://shopify/ProductVariant/19435458986123",
    "title":"52",
   }
],
....,
}
```